content/blog/2019-11-13-loop-perforation/index.md

```diff
@@ -0,0 +1,420 @@
+
+; <label>:4:
+  %5 = add nsw i32 %.01, %.0
+  br label %6
+
+; <label>:6:
+  %7 = add nsw i32 %.0, 2              ;; <- Perforated rate here
+  br label %2
+```
+After the driver perforates the desired loop, it executes that modified executable on the representative input.
+The driver checks the return code of the execution and treats crashing programs as unacceptable.
+It also measures the execution of each program in wall-clock time.
+The output is saved to disk then compared with the standard, non-perforated output.
+The comparison uses the application-specific `errors` module, calculating the error between the perforated and expected result for a number of user-defined metrics.
+This driver repeats this for every potential perforation rate, and then write the results out as another JSON file.
+
+For `sum_to_n`, this looks like the following (with some light post-processing for brevity):
+
+```
+{
+    "{\{\"sum_to_n\": {\"%2,%4,%6\": 1}}}": {
+        "return_code": 0,
+        "time": 0.0195,
+        "errors": {
+            "error_ratio": 0.0
+        }
+    },
+    "{\{\"sum_to_n\": {\"%2,%4,%6\": 2}}}": {
+        "return_code": 0,
+        "time": 0.0194,
+        "errors": {
+            "error_ratio": 0.4
+        }
+    },
+    "{\{\"sum_to_n\": {\"%2,%4,%6\": 3}}}": {
+        "return_code": 0,
+        "time": 0.0197,
+        "errors": {
+            "error_ratio": 0.7
+        }
+    },
+    "{\{\"sum_to_n\": {\"%2,%4,%6\": 5}}}": {
+        "return_code": 0,
+        "time": 0.0196,
+        "errors": {
+            "error_ratio": 1.0
+        }
+    }
+    // ...
+}
+```
+
+Here, our wall-clock execution time is essentially a wash, because the loop is so small that the difference is negligible relative to noise.
+However, we can see that the error ratio does drastically change as we increase the perforation rate, from perfect for the standard run to completely unacceptable at a perforation rate of 5, as we might expect (skipping 4 in every 5 iterations is skipping every integer except 0!) Note that if our application had multiple loops, we would see a results equal to the number of loops times the number of perforation rates (one run for each.)
+
+The final task of the driver is to combine what it learned about how much it can mangle each loop into one final answer.
+That is, where we were previously perforating each loop in turn, we now want to perforate some subset of the loops that we determined don't hurt the accuracy too much.
+Here, we follow the lead of the original loop perforated paper in making a greedy assumption—that we can choose an final loop perforation strategy based on joining the maximum perforation rate for each loop that that was below the error tolerance.
+This strategy makes the very optimistic assumption that errors are not additive; in the published paper their system backtracks in the case that the combined executable is unacceptable.
```

Comment:
Hmm, perhaps you mean that "errors _are_ additive"? That is, the error from perforating one loop and the error from perforating another get added together when you perforate them both?